ninth-realm/heimdall
# 9 api keys

## Summary

This PR adds clients and API keys to the system. A client is intended to be an authenticated party who does not have a username/password. Usually this will be a service. Clients are able to authenticate via API keys.

## Testing

### Clients

1. Create a client with
```
$ curl -X POST \
-d '{"name":"Test Client", "enabled":true}' \
'http://localhost:8080/api/v1/clients'
```
2. Retrieve a list of clients. Ensure the new client is in this list
```
$ curl -X GET 'http://localhost:8080/api/v1/clients'
```
3. Retrieve the new client.
```
$ curl -X GET 'http://localhost:8080/api/v1/clients/<ID FROM STEP 1>'
```
4. Update the client and retrieve to ensure the changes took place
```
$ curl -X PATCH \
-d '{"name":"Test Client Updated", "enabled":false}' \
'http://localhost:8080/api/v1/clients/<ID FROM STEP 1>'
$ curl -X GET 'http://localhost:8080/api/v1/clients/<ID FROM STEP 1>'
```
5. Delete the client
```
$ curl -X DELETE 'http://localhost:8080/api/v1/clients/<ID FROM STEP 1>'
$ curl -X GET 'http://localhost:8080/api/v1/clients/<ID FROM STEP 1>' # should return 404
```

### API Keys

Ensure you have a client created.

1. Generate an API key
```
$ curl -X POST \
-d '{"description":"My key"}' \
'http://localhost:8080/api/v1/clients/<CLIENT ID>/api-keys'
```
2. List all keys for a client and ensure the key is present
```
$ curl -X GET 'http://localhost:8080/api/v1/clients/<CLIENT ID>/api-keys'
```
3. Delete the key

```
$ curl -X GET 'http://localhost:8080/api/v1/clients/<CLIENT ID>/api-keys/<KEY ID>'
$ curl -X GET 'http://localhost:8080/api/v1/clients/<CLIENT ID>/api-keys' # should be empty
```
